src/field_theory/primitive_element.lean

```diff
@@ -0,0 +1,327 @@
+               (polynomial.X - polynomial.C (ϕ β))
+               (s.map (λ (a : E), polynomial.X - polynomial.C a))
+               (λ x hx, begin
+    rw multiset.mem_map at hx,
+    cases hx with a ha,
+    rw h_roots a at ha,
+    exact ha.2.symm,
+    rw polynomial.mem_roots (show polynomial.map ϕ h ≠ 0, by exact polynomial.map_ne_zero h_ne_zero),
+    cases multiset.exists_cons_of_mem ha.1 with y hy,
+    simp [hs, hy],
+  end),
+  rw [h', multiset.prod_repeat, multiset.card_map] at hs,
+  have hf : ¬is_unit (polynomial.X - polynomial.C (ϕ β)),
+  { rw [polynomial.is_unit_iff_degree_eq_zero,polynomial.degree_X_sub_C],
+    exact dec_trivial, },
+  have hn : s.card ≠ 0,
+  { intro hs_card,
+    rw [hs_card, pow_zero, mul_one, ←polynomial.map_C] at hs,
+    rw [polynomial.map_injective _ ϕ.injective hs, polynomial.eval_C, polynomial.leading_coeff_eq_zero] at h_root,
+    exact h_ne_zero h_root, },
+  have h_map_separable : (h.map ϕ).separable := polynomial.separable.map h_sep,
+  rw hs at h_map_separable,
+  rw [(polynomial.separable.of_pow hf hn (polynomial.separable.of_mul_right h_map_separable)).2, pow_one] at hs,
+  simp [hs],
+end
+
+end
+
+variables {F : Type*} [field F] {E : Type*} [field E] [algebra F E]
+
+lemma primitive_element_two_inf_key (α β : E) [F_sep : is_separable F E] (F_inf : infinite F) :
+  ∃ c : F, β ∈ F⟮α + (algebra_map F E) c * β⟯ :=
+begin
+  rcases F_sep α with ⟨hα, hf⟩,
+  rcases F_sep β with ⟨hβ, hg⟩,
+  let f := minimal_polynomial hα,
+  let g := minimal_polynomial hβ,
+  let ιFE := algebra_map F E,
+  let g' := g.map ιFE,
+  let E' := polynomial.splitting_field g',
+  let ιEE' := algebra_map E E',
+  let ιFE' := ιEE'.comp(ιFE),
+  have key := primitive_element_two_aux ιFE' (ιEE' α) (ιEE' β) f g,
+  cases key with c hc,
+  use c,
+  let γ := α+(ιFE c)*β,
+  change β ∈ F⟮γ⟯,
+  let f' := (f.map ιFE).comp(polynomial.C γ-(polynomial.C (ιFE c)) * (polynomial.X)),
+  let h := euclidean_domain.gcd f' g',
+  have h_sep : h.separable,
+  { cases euclidean_domain.gcd_dvd_right f' g' with p mul,
+    dsimp[←h] at mul,
+    apply polynomial.separable.of_mul_left,
+    rw ←mul,
+    exact polynomial.separable.map hg, },
+  have h_ne_zero : h ≠ 0,
+  { intro h_eq_zero,
+    rw euclidean_domain.gcd_eq_zero_iff at h_eq_zero,
+    apply polynomial.map_monic_ne_zero (minimal_polynomial.monic hβ) h_eq_zero.2, },
+  have h_root : h.eval β = 0,
+  { apply polynomial.gcd_eval_zero,
+    rw [polynomial.eval_comp, polynomial.eval_sub, polynomial.eval_mul],
+    rw [polynomial.eval_C, polynomial.eval_C, polynomial.eval_X, add_sub_cancel],
+    repeat { rw [polynomial.eval_map, ←polynomial.aeval_def, minimal_polynomial.aeval], }, },
+  have h_splits : polynomial.splits (algebra_map E E') h :=
+    polynomial.splits_of_splits_of_dvd (algebra_map E E')
+      (polynomial.map_ne_zero (minimal_polynomial.ne_zero hβ))
+      (polynomial.splitting_field.splits g') (euclidean_domain.gcd_dvd_right f' g'),
```

Comment:
Could you turn this into a separate lemma with the conclusion `splits _ (gcd f g)`?